collect_app/src/androidTest/java/org/odk/collect/android/preferences/qr/QrCodeActivitiesTest.java

```diff
@@ -0,0 +1,126 @@
+                    Manifest.permission.WRITE_EXTERNAL_STORAGE,
+                    Manifest.permission.READ_PHONE_STATE
+            ))
+            .around(new ResetStateRule(new AppDependencyModule() {
+                @Override
+                @Provides
+                public QRCodeGenerator providesQRCodeGenerator() {
+                    return new QRCodeGenerator() {
+                        @Override
+                        public Observable<Bitmap> generateQRCode(Collection<String> selectedPasswordKeys) {
+                            return Observable.create(emitter -> {
+                                Bitmap bitmap =
+                                        BitmapFactory.decodeResource(
+                                                ApplicationProvider.getApplicationContext().getResources(),
+                                                checkerBackgroundDrawableId);
+                                emitter.onNext(bitmap);
+                                emitter.onComplete();
+                            });
+                        }
+                    };
+                }
+            }));
+
+    @Before
+    public void stubAllExternalIntents() {
+        // By default Espresso Intents does not stub any Intents. Stubbing needs to be setup before
+        // every test run. In this case all external Intents will be blocked.
+        intending(not(isInternal())).respondWith(new Instrumentation.ActivityResult(Activity.RESULT_OK, null));
+    }
+
+    @Test
+    public void checkQRScannerIsInitiated() {
+        new MainMenuPage(rule)
+                .assertOnPage()
+                .clickOnMenu()
+                .clickConfigureQR()
+                .clickScanFragment()
+                .checkIsIdDisplayed(R.id.zxing_barcode_surface);
+    }
+
+    @Test
+    public void checkQRCodeImageViewDisplaysImage() {
+        new MainMenuPage(rule)
+                .assertOnPage()
+                .clickOnMenu()
+                .clickConfigureQR()
+                .clickViewQRFragment()
+                .assertImageViewShowsImage(R.id.ivQRcode, BitmapFactory.decodeResource(
+                        ApplicationProvider.getApplicationContext().getResources(),
+                        checkerBackgroundDrawableId
+                ));
+    }
+
+    @Test
+    public void pressImportQRCode() {
+        new MainMenuPage(rule)
+                .assertOnPage()
+                .clickOnMenu()
+                .clickConfigureQR()
+                .clickOnMenu()
+                .clickOnString(R.string.import_qrcode_sd);
+        
+        intended(hasAction(Intent.ACTION_PICK));
+        intended(hasType("image/*"));
+    }
+
+    @Test
+    public void pressShareQRCode() {
+        new MainMenuPage(rule)
+                .assertOnPage()
+                .clickOnMenu()
+                .clickConfigureQR()
+                .clickOnId(R.id.menu_item_share);
+
+        intended(hasAction(Intent.ACTION_CHOOSER));
```

Comment:
I think you still want another assertion here on the file that's sent right?

Reply:
yes, I'm not quite sure how though

Reply:
I think you want to check that the intent has the URI in it right?

Reply:
yes, but this Intent wraps another Intent with contains the URI as an Extra. I'm having trouble accessing this inner intent.

Reply:
nvm, figured it out! can you please take another look? I'm a little worried now that the test is too coupled with the code, especially the `expectedUri` part

https://github.com/hypercubestart/collect/blob/4005fe290c0d61bdd52271c16e0ce4a9f53cf4ba/collect_app/src/androidTest/java/org/odk/collect/android/preferences/qr/ConfigureWithQRCodeTest.java#L125
        Uri expected = FileProvider.getUriForFile(ApplicationProvider.getApplicationContext(),
                BuildConfig.APPLICATION_ID + ".provider",
                new File(path));

which I copied over directly